sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1938,8 +2080,48 @@ pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
   this->ZeEventList = nullptr;
   this->PiEventList = nullptr;
 
+  if (CurQueue->isInOrderQueue() && CurQueue->LastCommandEvent != nullptr) {
+    if (CurQueue->Device->useImmediateCommandLists()) {
+      if (ReuseDiscardedEvents && CurQueue->isDiscardEvents()) {
+        // If queue is in-order with discarded events and if
+        // new command list is different from the last used command list then
+        // signal new event from the last immediate command list. We are going
+        // to insert a barrier in the new command list waiting for that event.
+        auto QueueGroup = CurQueue->getQueueGroup(UseCopyEngine);
+        auto NextImmCmdList = QueueGroup.ImmCmdLists[QueueGroup.NextIndex];
+        if (CurQueue->LastCommandList != CurQueue->CommandListMap.end() &&
+            CurQueue->LastCommandList != NextImmCmdList) {
+          CurQueue->signalEventFromCmdListIfLastEventDiscarded(
+              CurQueue->LastCommandList);
+        }
+      }
+    } else {
+      // Ensure LastCommandEvent's batch is submitted if it is differrent
+      // from the one this command is going to.
```

Comment:
Please also comment that the "signalEventFromCmdListIfLastEventDiscarded" will be called at batch close if needed.

Reply:
Added a comment.